cmd/registration-dns/regforwarder.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"bytes"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"github.com/refraction-networking/gotapdance/pkg/dns-registrar/responder"
+	pb "github.com/refraction-networking/gotapdance/protobuf"
+	"google.golang.org/protobuf/proto"
+)
+
+// Interface to forward DNS registration requests. Use a dns responder to receive requests and send responses.
+type DnsRegForwarder struct {
+	// Endpoints to use in registration request
+	endpoint   string
+	bdendpoint string // bidirectional
+
+	// HTTP clients to use in request
+	client *http.Client
+
+	// dns responder to recieve and forward responses with
+	dnsResponder *responder.Responder
+}
+
+// Create new DnsRegForwarder.
+func NewDnsRegForwarder(endpoint string, bdendpoint string, dnsResponder *responder.Responder) (*DnsRegForwarder, error) {
+	f := DnsRegForwarder{}
+
+	httpTransport := http.DefaultTransport.(*http.Transport).Clone()
+	f.client = &http.Client{Transport: httpTransport}
+
+	f.endpoint = endpoint
+	f.bdendpoint = bdendpoint
+	f.dnsResponder = dnsResponder
+
+	return &f, nil
+}
+
+// Define function on how to respond to incoming dns requests and pass it to dnsResponder.
+func (f *DnsRegForwarder) RecvAndForward() error {
+	// send the raw request payload to the api endpoint and forward its response
+	forwardWith := func(reqIn []byte) ([]byte, error) {
+		regReq := &pb.C2SWrapper{}
+		err := proto.Unmarshal(reqIn, regReq)
+		if err != nil {
+			log.Printf("Error in request unmarshal: [%v]", err)
+			return nil, err
+		}
+
+		reqIsBd := regReq.GetRegistrationSource() == pb.RegistrationSource_BidirectionalDNS
+		if reqIsBd {
+			log.Println("Received bidirectional request")
```

Comment:
We should probably add a logging verbosity parameter. Right now it looks like each registration will generate 5+ log lines. Which is great for debug, but will likely overwhelm the system with log messages of limited use, generally we want to limit to 1 log line per registration. Either success and some details (regID, response code etc.) or error if one is encountered.

Reply:
I agree. I've made the logs use different log levels using the same `logrus` library as the client. The log level can be set using the `-loglevel` cli option or `log_level` in the config file. It should print no more than 1 log per request if the log level is set to `info`.